src/Config/MergeConfig.cs

```diff
@@ -0,0 +1,126 @@
+                {
+                    throw new InvalidOperationException($"The original JSON document to merge new content into must be a container type. Instead it is {root1.ValueKind}.");
+                }
+
+                if (root1.ValueKind != root2.ValueKind)
+                {
+                    return originalJson;
+                }
+
+                if (root1.ValueKind == JsonValueKind.Array)
+                {
+                    MergeArrays(jsonWriter, root2);
+                }
+                else
+                {
+                    MergeObjects(jsonWriter, root1, root2);
+                }
+            }
+
+            return Encoding.UTF8.GetString(outputBuffer.WrittenSpan);
+        }
+
+        /// <summary>
+        /// This helper methods tries to merge two Json Object recursively.
+        /// </summary>
+        private static void MergeObjects(Utf8JsonWriter jsonWriter, JsonElement root1, JsonElement root2)
+        {
+            jsonWriter.WriteStartObject();
+
+            // If a property exists in both documents, either:
+            // * Merge them, if the value kinds match (e.g. both are objects or arrays) recursively,
+            // * Completely override the value of the first with the one from the second, if the value kind mismatches (e.g. one is object, while the other is an array or string),
+            // * Or favor the value of the first, if the second one is null.
+            foreach (JsonProperty property in root1.EnumerateObject())
+            {
+                string propertyName = property.Name;
+
+                JsonValueKind newValueKind;
+
+                if (root2.TryGetProperty(propertyName, out JsonElement newValue) && (newValueKind = newValue.ValueKind) != JsonValueKind.Null)
+                {
+                    jsonWriter.WritePropertyName(propertyName);
+
+                    JsonElement originalValue = property.Value;
+                    JsonValueKind originalValueKind = originalValue.ValueKind;
+
+                    if (newValueKind == JsonValueKind.Object && originalValueKind == JsonValueKind.Object)
+                    {
+                        MergeObjects(jsonWriter, originalValue, newValue); // Recursive call
+                    }
+                    else if (newValueKind == JsonValueKind.Array && originalValueKind == JsonValueKind.Array)
+                    {
+                        MergeArrays(jsonWriter, newValue);
+                    }
+                    else
+                    {
+                        newValue.WriteTo(jsonWriter);
+                    }
+                }
+                else
+                {
+                    property.WriteTo(jsonWriter);
+                }
+            }
+
+            // Write all the properties of the second document that are unique to it.
+            foreach (JsonProperty property in root2.EnumerateObject())
+            {
+                if (!root1.TryGetProperty(property.Name, out _))
```

Comment:
Does this TryGetProperty check only for the immediate children of root1 or any property in the hierarchy upto the leaf?

What happens if `entities.book.rest` is present in root2 but in root1 `runtime.rest` is present. Will the above lookup suuceed?

Reply:
yes, it does a recursive merge.

Reply:
What is the answer to:

What happens if `entities.book.rest` is present in root2 but in root1 `runtime.rest` is present. Will the above lookup succeed?

If this lookup succeeds, the `entities.book.rest` from root2 will not be written to the merged config even though it should have been, Is this desired?

Reply:
the merged config will have both `runtime.rest` and `entities.book.rest`, since they have different parent. every object is merged separately